src/serialize/index.ts

```diff
@@ -66,46 +67,103 @@ export const deserialize = (input: string) =>
+
+export const printElement = (
+  el: Element | Document,
+  printColors = true,
+  depth = 3,
+) => {
   if (el instanceof Document) return '#document';
   let contents = '';
   const attrs = [...el.attributes];
   const splitAttrs = attrs.length > 2;
-  if (depth > 0 && el.childNodes.length <= 3) {
-    const singleLine =
-      !splitAttrs &&
-      (el.childNodes.length === 0 ||
-        (el.childNodes.length === 1 && el.childNodes[0] instanceof Text));
-    for (const child of el.childNodes) {
+  let needsMultipleLines = false;
+  if (depth > 0 && el.childNodes.length <= 5) {
+    const whiteSpaceSetting = getComputedStyle(el).whiteSpace;
+    const printedChildren: string[] = [];
+    let child = el.firstChild;
+    while (child) {
       if (child instanceof Element) {
-        contents += `\n  ${printElement(child, depth - 1).replace(
-          /\n/g,
-          '\n  ',
-        )}`;
+        needsMultipleLines = true;
+        printedChildren.push(printElement(child, printColors, depth - 1));
       } else if (child instanceof Text) {
-        contents += `${singleLine ? '' : '\n  '}${child.textContent}`;
+        // Merge consecutive text nodes together so their text can be collapsed
+        let consecutiveMergedText = child.textContent || '';
+        while (child.nextSibling instanceof Text) {
+          // We are collecting the consecutive siblings' text here
+          // so we are also skipping those siblings from being used by the outer loop
+          child = child.nextSibling;
+          consecutiveMergedText += child.textContent || '';
+        }
+        printedChildren.push(
+          whiteSpaceSetting === '' ||
+            whiteSpaceSetting === 'normal' ||
+            whiteSpaceSetting === 'nowrap' ||
+            whiteSpaceSetting === 'pre-line'
+            ? consecutiveMergedText.replace(
+                // Pre-line should collapse whitespace _except newlines
+                whiteSpaceSetting === 'pre-line' ? /[^\S\n]+/g : /\s+/g,
+                ' ',
+              )
+            : consecutiveMergedText,
+        );
       }
+      child = child.nextSibling;
     }
+    if (!needsMultipleLines)
+      needsMultipleLines =
+        splitAttrs || printedChildren.some((c) => c.includes('\n'));
 
-    if (!singleLine) contents += '\n';
+    contents += needsMultipleLines
+      ? `\n${printedChildren
+          .filter((c) => c.trim() !== '')
+          .map((c) => indent(c))
+          .join('\n')}\n`
+      : printedChildren.join('');
   } else {
     contents = '[...]';
   }
 
   const tagName = el.tagName.toLowerCase();
   const selfClosing = el.childNodes.length === 0;
-  return `<${tagName}${
+  // We haver to tell kolorist to print the colors
+  // beacuse by default it won't since we are in the browser
+  // (the colored message gets sent to node to be printed)
+  colors.options.enabled = true;
+  colors.options.supportLevel = 1;
+
+  // Syntax highlighting groups
+  const hi = {
```

Comment:
What does `hi` stand for? Can this be renamed to something more descriptive? 🤔

Reply:
It stands for highlight. I guess we can rename it to highlight :shrug: